### PR TITLE
bpo-20774: Add a JSON serializer to collections.deque

### PR DESCRIPTION
**Pull request by lisroach**
PR for [issue 20774](http://bugs.python.org/issue20774), adding a JSON serializer to collections.deque.

https://bugs.python.org/issue20774

### CODE REVIEW

**Review by rhettinger — CHANGES_REQUESTED** [review]
Over this looks good. Can you please add some tests.

File: Modules/_json.c [hunk 1]
```diff
@@ -1493,6 +1493,8 @@ encoder_listencode_obj(PyEncoderObject *s, _PyAccu *acc,
     /* Encode Python object obj to a JSON term */
     PyObject *newobj;
     int rv;
+    PyObject *mo = PyImport_ImportModule("collections");
```
Comment (Birne94) [inline Modules/_json.c:1496]: Wouldn't you need to decrease the reference counter to `mo` and `deque_type` after using them?

**Comment by brettcannon** [comment]
To try and help move older pull requests forward, we are going through and backfilling 'awaiting' labels on pull requests that are lacking the label. Based on the current reviews, the best we can tell in an automated fashion is that a core developer requested changes to be made to this pull request.

If/when the requested changes have been made, please leave a comment that says, `I have made the requested changes; please review again`. That will trigger a bot to flag this pull request as ready for a follow-up review.

**Comment by sakurai-youhei** [comment]
For those who arrived here like me when searching quick solution to encode dqeue object into JSON; Below code (as per [documentation](https://docs.python.org/3/library/json.html#json.JSONEncoder.default)) shall be good enough.

```
from collections import deque
from json import dumps, JSONEncoder

class JSONEncoderSupportingArbitraryIterators(JSONEncoder):
    def default(self, o):
        try:
            iterable = iter(o)
        except TypeError:
            pass
        else:
            return list(iterable)
        # Let the base class default method raise the TypeError
        return json.JSONEncoder.default(self, o)

dumps(deque(range(10)), cls=JSONEncoderSupportingArbitraryIterators)
```

**Comment by carlbordum** [comment]
@lisroach are you interested in writting the tests, or may I? :)

**Comment by lisroach** [comment]
Sorry, just realized I had this sitting for so long! @carlbordum feel free to add more tests if you think of them :)

File: Lib/test/test_json/test_dump.py [hunk 2]
```diff
@@ -3,6 +3,7 @@
 
 from test.support import bigmemtest, _1G
 
+
```
Comment (matrixise) [inline Lib/test/test_json/test_dump.py]: Maybe you could remove this blank line

File: Lib/test/test_json/test_dump.py [hunk 3]
```diff
@@ -47,7 +48,6 @@ def __lt__(self, o):
         d[1337] = "true.dat"
         self.assertEqual(self.dumps(d, sort_keys=True), '{"1337": "true.dat"}')
```
Comment (matrixise) [inline Lib/test/test_json/test_dump.py]: Why this line ?

**Comment by csabella** [comment]
@lisroach, should this pull request be closed based on the performance benchmarks mentioned on the bug tracker or were you still working on this change?  Thanks!

**Comment by erlend-aasland** [comment]
@lisroach, are you going to follow up this PR?

**Comment by kumaraditya303** [comment]
Closing because of lack of response and conflicts, feel free to reopen if you want to work on this.

**Pull request closed without merging**